README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# cleanventory <img src="man/figures/logo.svg" align="right" height="139" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/RaoulWolf/cleanventory/workflows/R-CMD-check/badge.svg)](https://github.com/RaoulWolf/cleanventory/actions)
[![Codecov test
coverage](https://codecov.io/gh/RaoulWolf/cleanventory/branch/master/graph/badge.svg)](https://app.codecov.io/gh/RaoulWolf/cleanventory?branch=master)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

> A [ZeroPM](https://zeropm.eu/) R package

The goal of cleanventory is to provide simple functionality to clean and
partially curate data sets of common chemical inventories. The aim is to
document every step, from the raw (downloaded) files to the final
tables.

cleanventory aims to correctly identify all missing values in data sets,
validates CAS Registry Numbers (when present) and additionally offers
functionality to transform all special characters into ASCII characters.

The dependencies of cleanventory are kept at as minimal as possible:
[openxlsx](https://cran.r-project.org/web/packages/openxlsx) for
handling .xlsx files, and the trio of
[pdftools](https://cran.r-project.org/web/packages/pdftools),
[magick](https://cran.r-project.org/web/packages/magick) and
[tesseract](https://cran.r-project.org/web/packages/tesseract) to
extract data from (image) .pdf files.

We suggest the following packages/functionalities in addition:
[`bit64::as.integer64()`](https://cran.r-project.org/web/packages/bit64)
to correctly handle the `us_tsca$cas_reg_no` and
`us_cdr$chemical_id_wo_dashes` columns (kept as `double` for
compatibility).

As of 2022-08-02, the following inventories are included:

| Chemical Inventory | Function          | Compatible Version(s)                        | URL                                                                                                                                                     |
|:-------------------|:------------------|:---------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------|
| US TSCA            | `read_us_tsca()`  | 2021-08                                      | <https://www.epa.gov/tsca-inventory>                                                                                                                    |
| EU CLP Annex VI    | `read_eu_clp()`   | 9, 10, 13, 14, 15, 17                        | <https://echa.europa.eu/en/information-on-chemicals/annex-vi-to-clp>                                                                                    |
| EU ECI             | `read_eu_eci()`   | *Unknown*                                    | <https://echa.europa.eu/information-on-chemicals/ec-inventory>                                                                                          |
| Japan NITE         | `read_jp_nite()`  | March 2022                                   | <https://www.nite.go.jp/chem/english/ghs/ghs_download.html>                                                                                             |
| New Zealand IoC    | `read_nz_ioc()`   | December 2021                                | <https://www.epa.govt.nz/database-search/new-zealand-inventory-of-chemicals-nzioc/>                                                                     |
| South Korea NCIS   | `read_kr_ncis()`  | 4 May 2022                                   | <https://ncis.nier.go.kr/en/mttrList.do>                                                                                                                |
| Australia HCIS     | `read_au_hcis()`  | *Unknown*                                    | <http://hcis.safeworkaustralia.gov.au/HazardousChemical>                                                                                                |
| Australia ICI      | `read_au_ici()`   | 10 February 2022                             | <https://www.industrialchemicals.gov.au/search-inventory>                                                                                               |
| Taiwan CSI         | `read_tw_csi()`   | *Unknown*                                    | <https://gazette.nat.gov.tw/egFront/detail.do?metaid=73440&log=detailLog></br><https://gazette.nat.gov.tw/egFront/detail.do?metaid=78617&log=detailLog> |
| Philippine ICCS    | `read_ph_iccs()`  | 2017, 2020, 2021                             | <https://chemical.emb.gov.ph/?page_id=138>                                                                                                              |
| Japan CSCL         | `read_jp_cscl()`  | 31 May 2022</br>31 May 2022</br>1 April 2022 | <https://www.nite.go.jp/en/chem/chrip/chrip_search/sltLst>                                                                                              |
| Canada DSL         | `read_ca_dsl()`   | 14 June 2022                                 | <https://pollution-waste.canada.ca/substances-search/Substance?lang=en>                                                                                 |
| China IECSC        | `read_cn_iecsc()` | 2013                                         | <https://www.mee.gov.cn/gkml/hbb/bgg/201301/t20130131_245810.htm>                                                                                       |
| Nordics SPIN       | `read_xn_spin()`  | 2000                                         | <http://www.spin2000.net/spinmyphp/>                                                                                                                    |
| US CDR             | `read_us_cdr()`   | 2016</br>2020                                | <https://www.epa.gov/chemical-data-reporting>                                                                                                           |
| Malaysia CIMS      | `read_my_cims()`  | 2017                                         | <https://cims.dosh.gov.my/>                                                                                                                             |

## Installation

You can install the development version of cleanventory from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
remotes::install_github("ZeroPM-H2020/cleanventory")
```

## Examples

This is an example which shows you how to get the data set of the
(current) EU CLP Annex VI:

``` r
library(cleanventory)

tmp <- tempdir()

url <- paste0(
  "https://echa.europa.eu/documents/10162/17218/",
  "annex_vi_clp_table_atp17_en.xlsx/",
  "4dcec79c-f277-ed68-5e1b-d435900dbe34?t=1638888918944"
)

eu_clp_file <- download.file(
  url, 
  destfile = paste(tmp, "annex_vi_clp_table_atp17_en.xlsx", sep = "/"),
  quiet = TRUE,
  mode = ifelse(.Platform$OS.type == "windows", "wb", "w")
)

path <- paste(tmp, "annex_vi_clp_table_atp17_en.xlsx", sep = "/")

eu_clp <- read_eu_clp(path)

invisible(file.remove(path))

head(eu_clp)
#>       index_no international_chemical_identification     ec_no     cas_no
#> 1 001-001-00-9                              hydrogen 215-605-7  1333-74-0
#> 2 001-002-00-4             aluminium lithium hydride 240-877-9 16853-85-3
#> 3 001-003-00-X                        sodium hydride 231-587-3  7646-69-7
#> 4 001-004-00-5                       calcium hydride 232-189-2  7789-78-8
#> 5 003-001-00-4                               lithium 231-102-5  7439-93-2
#> 6 003-002-00-X                        n-hexyllithium 404-950-0 21369-64-2

str(eu_clp)
#> 'data.frame':    4702 obs. of  4 variables:
#>  $ index_no                             : chr  "001-001-00-9" "001-002-00-4" "001-003-00-X" "001-004-00-5" ...
#>  $ international_chemical_identification: chr  "hydrogen" "aluminium lithium hydride" "sodium hydride" "calcium hydride" ...
#>  $ ec_no                                : chr  "215-605-7" "240-877-9" "231-587-3" "232-189-2" ...
#>  $ cas_no                               : chr  "1333-74-0" "16853-85-3" "7646-69-7" "7789-78-8" ...
```

## Acknowledgement

This R package was developed at the [Norwegian Geotechnical Institute
(NGI)](https://www.ngi.no/eng) as part of the project [ZeroPM: Zero
pollution of Persistent, Mobile substances](https://zeropm.eu/). This
project has received funding from the European Union’s Horizon 2020
research and innovation programme under grant agreement No 101036756.

------------------------------------------------------------------------

If you find this package useful and can afford it, please consider
making a donation to a humanitarian non-profit organization, such as
[Sea-Watch](https://sea-watch.org/en/). Thank you.


NEWS.md
# cleanventory 0.3.2 (2022-09-14) 

* Added new functions for the Malaysia Chemical Information Management System 
  (`read_my_cims()`), the United States Chemical Data Reporting Inventory 
  (`read_us_cdr()`), and the Substances in Preparations in Nordic Countries 
  inventory (`read_xn_spin()`). 

# cleanventory 0.3.1 (2022-07-19)

* Fixed function to clean up non-ASCII characters. 
* Added new function for the Chinese Inventory of Existing Chemical Substances 
  and Chemicals (`read_cn_iecsc()`). 
* Fixed tests. 

# cleanventory 0.3.0 (2022-07-15)

* Added 2-letter country/sovereign region indication to each function to make it
  easier to choose the right function. For example, `read_clp()` has become
  `read_eu_clp()`.
* Added new functions for new inventories: the Taiwan Chemical Substances 
  Inventory (`read_tw_csi()`), the Philippine Inventory of Chemicals and 
  Chemical Substances (`read_ph_iccs()`), the Japan Chemical Substances Control 
  Law (`read_jp_cscl()`), the Australia Inventory of Industrial Chemicals 
  (`read_au_iic()`), and the Canada Domestic Substances List (`read_ca_dsl()`).
* The tests have been updated to reflect the name changes of the functions. 

# cleanventory 0.2.4 (2022-05-30)

* Added functionality to read the Australia Industrial Chemicals Inventory 
  (`read_ici()`). 
* Added function `date_to_cas()` to repair the well-known .xls(x) 
  malformatting.
* Updated internal function `.clean_non_ascii()` with more characters. 

# cleanventory 0.2.3 (2022-05-04)

* Added functionality to read the South Korea National Chemicals Information 
  System (`read_ncis()`).
* Added functionality to read the Australia Hazardous Chemical Information 
  System (`read_hcis()`).
* Added functionality to reasonably clean non-ASCII characters 
  (`.clean_non_ascii()`).
* Detection of "date-ifyed" CAS Registry Numbers and automatic conversion back
  to CAS RNs for `read_ec()`.

# cleanventory 0.2.2 (2022-04-27)

* Adding functionality to read and clean the New Zealand Inventory of 
  Chemicals (`read_ioc()`).
* Fixed minor typos.

# cleanventory 0.2.1 (2022-04-26)

* Adding functionality to read and clean the Japanese NITE inventory.
* Updated information for the US EPA TSCA inventory. 
* Fixed rare timeout issue in testing

# cleanventory 0.1.0 (2022-01-19)

* Fixed Unix issue with `read_clp()`.

# cleanventory 0.0.4 (2022-01-14)

* Outsourcing data sets into seperate cleanventory.data package.
* Changing dependency from readxl to openxlsx.
* Fixing tests to reflect the above changes. 
* Adding new logo.

# cleanventory 0.0.3 (2022-01-11)

* Added ECHA CLP Annex VI data sets as `clp`.
* Added ECHA EC data set as `ec`.
* Added functions to clean original files: `clean_tsca()`, `clean_clp()` and 
  `clean_ec()`.

# cleanventory 0.0.1 (2022-01-07)

* Added US EPA TSCA data set (v. 2021-08) as `tsca`.
* Added support for tests *via* {tinytest}.
* Added support for continuous integration *via* GitHub Actions.
* Added support for code coverage *via* {covr}.
* Added a `NEWS.md` file to track changes to the package.


R/read_clp.R
#' @title Read-In and Clean the ECHA CLP Annex VI Data Set
#' @description This function reads-in and automatically cleans the ECHA CLP
#'   Annex VI data set.
#' @param path (Character) The path to the XLSX file.
#' @param clean_non_ascii (Logical) Should the non-ASCII characters be
#'   reasonably converted? Defaults to \code{FALSE}.
#' @param atp (Logical) Should the "atp" version information be included?
#'   Defaults to \code{TRUE}.
#' @details The function reads-in and cleans the ECHA CLP Annex VI data set
#'   into long flat format.
#' @return Returns a data frame.
#' @author Raoul Wolf (\url{https://github.com/RaoulWolf/})
#' @note Tested with "atp" versions 9, 10, 13, 14, 15, and 17.
#' @examples \dontrun{
#' download.file(
#'   url = paste0(
#'     "https://echa.europa.eu/documents/10162/17218/",
#'     "annex_vi_clp_table_atp17_en.xlsx/",
#'     "4dcec79c-f277-ed68-5e1b-d435900dbe34?t=1638888918944"
#'   ),
#'   destfile = "annex_vi_clp_table_atp17_en.xlsx"
#' )
#'
#' path <- "annex_vi_clp_table_atp17_en.xlsx"
#'
#' clp <- read_clp(path)
#' }
#' @importFrom openxlsx read.xlsx
#' @export
read_clp <- function(path, clean_non_ascii = FALSE, atp = TRUE) {

  if (!is.logical(clean_non_ascii) || is.na(clean_non_ascii)) {
    clean_non_ascii <- FALSE
  }

  if (!is.logical(atp) || is.na(atp)) {
    atp <- TRUE
  }

  clp <- openxlsx::read.xlsx(
    xlsxFile = path,
    startRow = 7,
    colNames = FALSE,
    cols = 1:4,
    na.strings = "-"
  )

  colnames(clp) <- c(
    "index_no", "international_chemical_identification", "ec_no", "cas_no"
  )

  clp_split <- split(clp, clp$index_no)


  if (.Platform$OS.type == "windows") {
    pattern <- "\\[[[:digit:]]+\\]\n"
    mini_pattern <- "\n"
  } else {
    pattern <- "\\[[[:digit:]]+\\]\r\n"
    mini_pattern <- "\r\n"
  }

  clp_res <- lapply(
    clp_split,
    FUN = function(x) {

      if (any(grepl(pattern = pattern,
                    x$international_chemical_identification),
              grepl(pattern = pattern, x$ec_no),
              grepl(pattern = pattern, x$cas_no))) {

        international_chemical_identification_split <- unlist(
          strsplit(
            x$international_chemical_identification,
            split = pattern
          )
        )

        international_chemical_identification_index <- unlist(
          regmatches(
            x$international_chemical_identification,
            m = gregexpr(
              pattern = "\\[[[:digit:]]+\\]",
              text = x$international_chemical_identification
            )
          )
        )

        international_chemical_identification_index <- as.integer(
          unlist(
            regmatches(
              international_chemical_identification_index,
              m = gregexpr(
                pattern = "[[:digit:]]+",
                text = international_chemical_identification_index
              )
            )
          )
        )

        international_chemical_identification_split <- trimws(
          gsub(
            pattern = mini_pattern,
            replacement = " ",
            international_chemical_identification_split
          )
        )

        international_chemical_identification_df <- data.frame(
          index = international_chemical_identification_index,
          international_chemical_identification =
            international_chemical_identification_split,
          stringsAsFactors = FALSE
        )

        ec_no_split <- unlist(
          strsplit(x$ec_no, split = pattern)
        )

        ec_no_index <- unlist(
          regmatches(
            x$ec_no,
            m = gregexpr(pattern = "\\[[[:digit:]]+\\]", text = x$ec_no)
          )
        )

        ec_no_index <- as.integer(
          unlist(
            regmatches(
              ec_no_index,
              m = gregexpr(pattern = "[[:digit:]]+", text = ec_no_index)
            )
          )
        )

        if (length(ec_no_index) == 0L) {

          ec_no_index <- NA_integer_

        }

        ec_no_split <- trimws(
          gsub(
            pattern = paste0(mini_pattern, "]"),
            replacement = "",
            ec_no_split
          )
        )

        ec_no_split <- ifelse(
          test = nchar(ec_no_split) < 9L, yes = NA_character_, no = ec_no_split
        )

        ec_no_df <- data.frame(
          index = ec_no_index,
          ec_no = ec_no_split,
          stringsAsFactors = FALSE
        )

        cas_no_split <- unlist(
          strsplit(x$cas_no, split = pattern)
        )

        cas_no_index <- unlist(
          regmatches(
            x$cas_no,
            m = gregexpr(pattern = "\\[[[:digit:]]+\\]", text = x$cas_no)
          )
        )

        cas_no_index <- as.integer(
          unlist(
            regmatches(
              cas_no_index,
              m = gregexpr(pattern = "[[:digit:]]+", text = cas_no_index)
            )
          )
        )

        if (length(cas_no_index) == 0L) {

          cas_no_index <- NA_integer_

        }

        cas_no_split <- trimws(
          gsub(
            pattern = paste0(mini_pattern, "]"),
            replacement = "",
            cas_no_split
          )
        )

        cas_no_df <- data.frame(
          index = cas_no_index,
          cas_no = cas_no_split,
          stringsAsFactors = FALSE
        )

        final_df <- data.frame(
          index = 1:max(
            international_chemical_identification_index,
            ec_no_index,
            cas_no_index,
            na.rm = TRUE
          ),
          index_no = x$index_no,
          stringsAsFactors = FALSE
        )

        final_df <- merge(
          merge(
            merge(
              final_df,
              international_chemical_identification_df,
              by = "index",
              all.x = TRUE
            ),
            ec_no_df,
            by = "index",
            all.x = TRUE
          ),
          cas_no_df,
          by = "index",
          all.x = TRUE
        )

        final_df <- subset(final_df, select = -index)

      } else {

        final_df <- data.frame(
          index_no = x$index_no,
          international_chemical_identification = trimws(
            gsub(
              pattern = mini_pattern,
              replacement = " ", x$international_chemical_identification
            )
          ),
          ec_no = x$ec_no,
          cas_no = x$cas_no,
          stringsAsFactors = FALSE
        )

      }

      final_df$ec_no <- ifelse(
        test = final_df$index_no == "606-144-00-6" &
          final_df$cas_no == "57960-19-7",
        yes = NA_character_,
        no = final_df$ec_no
      )

      final_df

    }
  )

  clp_res <- do.call(what = "rbind", args = c(clp_res, make.row.names = FALSE))

  if (clean_non_ascii) {
    clp_res <- transform(
      clp_res,
      international_chemical_identification =
        .clean_non_ascii(international_chemical_identification)
    )
  }

  if (atp) {

    file_name <- unlist(strsplit(path, split = "/"))
    file_name <- file_name[grepl(pattern = ".xlsx", file_name)]

    atp_version <- regmatches(
      file_name,
      m = gregexpr(pattern = "[[:digit:]]", text = file_name)
    )
    atp_version <- as.integer(paste(unlist(atp_version), collapse = ""))

    clp_res <- transform(clp_res, atp = atp_version)

  }

  clp_res

}
